File: src/domain/datasusCatalog.ts
export type CatalogCode =
  | "SIH"
  | "SIA"
  | "CNES"
  | "SIM"
  | "SINASC"
  | "SINAN"
  | "PNI"
  | "ESUS_APS"
  | "CIHA";

export type CatalogDefinition = {
  code: CatalogCode;
  label: string;
};

export const FIRST_AVAILABLE_YEAR = 0;

export const DATASUS_CATALOGS: readonly CatalogDefinition[] = [
  { code: "SIH", label: "Hospital Information System" },
  { code: "SIA", label: "Ambulatory Information System" },
  { code: "CNES", label: "National Registry of Health Facilities" },
  { code: "SIM", label: "Mortality Information System" },
  { code: "SINASC", label: "Live Birth Information System" },
  { code: "SINAN", label: "Notifiable Diseases Information System" },
  { code: "PNI", label: "National Immunization Program" },
  { code: "ESUS_APS", label: "e-SUS / APS" },
  { code: "CIHA", label: "Hospitalization Communication System" },
];

export function currentYear(now: Date = new Date()): number {
  return now.getUTCFullYear();
}

export function availableYears(now: Date = new Date()): number[] {
  const endYear = currentYear(now);
  const years: number[] = [];
  for (let year = FIRST_AVAILABLE_YEAR; year <= endYear; year += 1) {
    years.push(year);
  }
  return years;
}

export function availableMonths(): number[] {
  return [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];
}

export function isKnownCatalog(code: string): code is CatalogCode {
  return DATASUS_CATALOGS.some((catalog) => catalog.code === code);
}

export function yearSelectionKey(catalog: CatalogCode, year: number): string {
  return `${catalog}:${String(year).padStart(4, "0")}`;
}

export function monthSelectionKey(
  catalog: CatalogCode,
  year: number,
  month: number,
): string {
  return `${catalog}:${String(year).padStart(4, "0")}:${String(month).padStart(2, "0")}`;
}

File: internal/domain/datasus_catalog.go
package domain

import (
	"fmt"
	"slices"
	"time"
)

type CatalogCode string

type CatalogDefinition struct {
	Code  CatalogCode
	Label string
}

const (
	CatalogSIH     CatalogCode = "SIH"
	CatalogSIA     CatalogCode = "SIA"
	CatalogCNES    CatalogCode = "CNES"
	CatalogSIM     CatalogCode = "SIM"
	CatalogSINASC  CatalogCode = "SINASC"
	CatalogSINAN   CatalogCode = "SINAN"
	CatalogPNI     CatalogCode = "PNI"
	CatalogESUSAPS CatalogCode = "ESUS_APS"
	CatalogCIHA    CatalogCode = "CIHA"
)

var KnownCatalogs = []CatalogDefinition{
	{Code: CatalogSIH, Label: "Hospital Information System"},
	{Code: CatalogSIA, Label: "Ambulatory Information System"},
	{Code: CatalogCNES, Label: "National Registry of Health Facilities"},
	{Code: CatalogSIM, Label: "Mortality Information System"},
	{Code: CatalogSINASC, Label: "Live Birth Information System"},
	{Code: CatalogSINAN, Label: "Notifiable Diseases Information System"},
	{Code: CatalogPNI, Label: "National Immunization Program"},
	{Code: CatalogESUSAPS, Label: "e-SUS / APS"},
	{Code: CatalogCIHA, Label: "Hospitalization Communication System"},
}

func CurrentYear() int {
	return time.Now().Year()
}

func CatalogCodes() []CatalogCode {
	out := make([]CatalogCode, 0, len(KnownCatalogs))
	for _, catalog := range KnownCatalogs {
		out = append(out, catalog.Code)
	}
	return out
}

func IsKnownCatalog(code CatalogCode) bool {
	for _, catalog := range KnownCatalogs {
		if catalog.Code == code {
			return true
		}
	}
	return false
}

func YearRange(startYear int, endYear int) []int {
	if endYear < startYear {
		return nil
	}
	years := make([]int, 0, endYear-startYear+1)
	for year := startYear; year <= endYear; year++ {
		years = append(years, year)
	}
	return years
}

func AvailableYears() []int {
	return YearRange(0, CurrentYear())
}

func AvailableMonths() []int {
	months := make([]int, 12)
	for i := range 12 {
		months[i] = i + 1
	}
	return months
}

func YearSelectionKey(catalog CatalogCode, year int) string {
	return fmt.Sprintf("%s:%04d", catalog, year)
}

func MonthSelectionKey(catalog CatalogCode, year int, month int) string {
	return fmt.Sprintf("%s:%04d:%02d", catalog, year, month)
}

func SortCatalogCodes(codes []CatalogCode) {
	slices.Sort(codes)
}
